libasn1compiler/asn1c_C.c

```diff
@@ -820,24 +820,23 @@ asn1c_lang_C_type_SEx_OF(arg_t *arg) {
 			&& memb->expr_type == ASN_BASIC_INTEGER))
 	    	&& expr_elements_count(arg, memb))) {
 		arg_t tmp;
-		asn1p_expr_t tmp_memb;
+		asn1p_expr_t *tmp_memb = memb;
```

Comment:
I don't understand the logic of this. You rewrite the existing object instead of mostly making a copy. Explain please?

Reply:
Inside function `register_global_name()`, the `address` of the `asn1p_expr_t', i.e. `arg->expr` 's value is used as an indication whether this expression was added into `used_names` or not. 

```
if(!(arg->expr->_mark & TM_NAMEGIVEN) && arg->expr != n->expr) {
```

For original code, the 
```
asn1p_expr_t tmp_memb;
tmp.expr = &tmp_memb;
```
The `tmp.expr` is actually pointed to an address of an auto-variable which resides inside stack.
Moreover, after we add it to `used_names`,

```
    n = calloc(1, sizeof(*n));
    assert(n);
    n->expr = arg->expr;   <-- this pointer points to an auto variable which no longer valid in other functions 
    n->name = strdup(name);
    TQ_ADD(&used_names, n, next);
```
it will be a nightmare if we want to free it or just only try to access it in other functions, e.g. de-reference it with `n->expr` in function `c_name_clash()` line `FATAL(.......)`.